src/util/invariant.h

```diff
@@ -239,84 +245,170 @@ invariant_violated_string(
     file, function, line, condition, reason);
 }
 
+namespace detail
+{
+template <typename T>
+struct always_falset : public std::false_type
+{
+};
+} // namespace detail
+
+/// Helper to give us some diagnostic in a usable form on assertion failure.
+/// For now only provides string output
+/// Must be specialised for all types that should be useable as diagnostics
+template <typename T>
+struct diagnostics_helpert
+{
+  // silly construct because C++ won't let us write static_assert(false,...)
+  static_assert(
+    detail::always_falset<T>::value,
+    "to avoid unintended strangeness, diagnostics_helpert needs to be "
+    "specialised for each diagnostic type");
+  static std::string diagnostics_as_string(const T &);
+};
+
+// Standard string specialisations for diagnostics_helper
+// (character arrays, character pointers and std::string)
+
+template <std::size_t N>
+struct diagnostics_helpert<char[N]>
+{
+  static std::string diagnostics_as_string(const char (&string)[N])
+  {
+    return string;
+  }
+};
+template <>
+struct diagnostics_helpert<char *>
+{
+  static std::string diagnostics_as_string(const char *string)
+  {
+    return string;
+  }
+};
+
+template <>
+struct diagnostics_helpert<std::string>
+{
+  // This is deliberately taking a copy instead of a reference
+  // to avoid making an extra copy when passing a temporary string
+  // as a diagnostic
+  static std::string diagnostics_as_string(std::string string)
+  {
+    return string;
+  }
+};
+
+namespace detail
+{
+inline std::string assemble_diagnostics()
+{
+  return "";
+}
+
+template <typename T>
+std::string diagnostic_as_string(T &&val)
+{
+  return diagnostics_helpert<
+    typename std::remove_cv<typename std::remove_reference<T>::type>::type>::
+    diagnostics_as_string(std::forward<T>(val));
+}
+
+inline void write_rest_diagnostics(std::ostream &)
+{
+}
+
+template <typename T, typename... Ts>
+void write_rest_diagnostics(std::ostream &out, T &&next, Ts &&... rest)
+{
+  out << ", " << diagnostic_as_string(std::forward<T>(next));
+  write_rest_diagnostics(out, std::forward<Ts>(rest)...);
+}
+
+template <typename T, typename... Ts>
+std::string assemble_diagnostics(T &&first, Ts &&... rest)
+{
+  std::ostringstream out;
+  out << "[ " << diagnostic_as_string(std::forward<T>(first));
+  write_rest_diagnostics(out, std::forward<Ts>(rest)...);
+  out << " ]";
```

Comment:
Can we separate the diagnostics by newlines? Some of them are expected to go over several lines (like `expr.pretty()`).

Reply:
Good point, I'd still want to bracket them somehow - how about

```
<< EXTRA DIAGNOSTICS >>
D1
D2
...
DN
<< END EXTRA DIAGNOSTICS>>
```

?